src/ripple/app/data/tests/SociDB.test.cpp

```diff
@@ -295,11 +295,77 @@ class SociDB_test final : public TestSuite
                 remove (dbPath);
         }
     }
+    void testSQLiteDeleteWithSubselect()
+    {
+        testcase ("deleteWithSubselect");
+        BasicConfig c;
+        setupSQLiteConfig (c, getDatabasePath ());
+        SociConfig sc (c, "SociTestDB");
+        soci::session s;
+        sc.open (s);
+        const char* dbInit[] = {
+            "BEGIN TRANSACTION;",
+            "CREATE TABLE Ledgers (                         \
+                LedgerHash      CHARACTER(64) PRIMARY KEY,  \
+                LedgerSeq       BIGINT UNSIGNED             \
+            );",
+            "CREATE INDEX SeqLedger ON Ledgers(LedgerSeq);",
+
+            "CREATE TABLE Validations   (                   \
+                LedgerHash  CHARACTER(64)                   \
+            );",
+            "CREATE INDEX ValidationsByHash ON              \
+                Validations(LedgerHash);",
+            "END TRANSACTION;"};
+        int dbInitCount = std::extent<decltype(dbInit)>::value;
+        for (int i = 0; i < dbInitCount; ++i)
+        {
+            s << dbInit[i];
+        }
+        char lh[65];
+        memset (lh, 'a', 64);
+        lh[64] = '\0';
+        int toIncIndex = 63;
+        int const numRows = 16;
+        for (int i = 0; i < numRows; ++i)
+        {
+            ++lh[toIncIndex];
+            if (lh[toIncIndex] == 'z')
+                --toIncIndex;
+            std::string slh(lh);
+            s << "INSERT INTO Ledgers (LedgerHash, LedgerSeq) VALUES "
+                 "(:lh, :li);", soci::use (slh), soci::use (i);
+            s << "INSERT INTO Validations (LedgerHash) VALUES "
+                 "(:lh);", soci::use (slh);
+        }
+        std::vector<int> ledgersLS (numRows * 2);
+        std::vector<std::string> validationsLH (numRows * 2);
+        s << "SELECT LedgerSeq FROM Ledgers;", soci::into (ledgersLS);
+        s << "SELECT LedgerHash FROM Validations;", soci::into (validationsLH);
+        expect (ledgersLS.size () == numRows &&
+                validationsLH.size () == numRows);
+        s << "DELETE FROM Validations WHERE LedgerHash IN "
+             "(SELECT Ledgers.LedgerHash FROM Validations JOIN Ledgers ON "
+             "Validations.LedgerHash=Ledgers.LedgerHash WHERE Ledgers.LedgerSeq < :num);",
+            soci::use (numRows / 2);
+        validationsLH.resize (numRows * 2);
+        s << "SELECT LedgerHash FROM Validations;", soci::into (validationsLH);
+        expect (validationsLH.size () == numRows / 2);
+
+        {
+            using namespace boost::filesystem;
+            // Remove the database
+            path dbPath (sc.connectionString ());
+            if (is_regular_file (dbPath))
+                remove (dbPath);
```

Comment:
The test looks great, though I'll add a couple of small suggestions after this. Unfortunately, it fails here b/c Windows takes locks on the database files while the session is open. This can be fixed by having the `soci::session s;` (currently line 304) in a separate scope that ends before this one begins.